code/modules/telesci/dses.dm

```diff
@@ -0,0 +1,291 @@
+				t += "<BR><A href='?src=\ref[src];eject=[current_module_num]'>Eject Module</A>"
+		else
+			t += "<BR>No modules detected"
+		t += "<BR>Module slots [module_list.len]/[module_limit]"
+		t += "<BR><A href='?src=\ref[src];eject_cell=1'>Eject Cell</A>"
+	else
+		t += "<B>No cell detected.</B>"
+	var/datum/browser/popup = new(user, "DSES", name, 300, 450)
+	popup.set_content(t)
+	popup.set_title_image(user.browse_rsc_icon(src.icon, src.icon_state))
+	popup.open()
+
+/obj/item/device/dses/Topic(href, href_list)
+	..()
+	usr.set_machine(src)
+	if(href_list["pulse"])
+		if (usr.get_active_hand() != src || usr.stat)
+			to_chat(usr, "<span class = 'caution'>You need to have \the [src] in your hand to do that!</span>")
+			return
+		if(C.charge < pulse_cost)
+			to_chat(usr, "<span class = 'caution'>\The [C] has insufficient charge to support another pulse.</span>")
+			return
+		handle_pulse(usr)
+
+	else if(href_list["eject"])
+		var/index = text2num(href_list["eject"])
+		if(index && index <= module_list.len)
+			var/obj/item/weapon/dses_module/D = module_list[index]
+			uninstall_module(D, usr)
+			D = null
+	else if(href_list["wipe"])
+		var/index = text2num(href_list["wipe"])
+		if(index && index <= positive_locations.len)
+			var/atom/A = positive_locations[index]
+			positive_locations -= A
+		else
+			positive_locations = list()
+	else if(href_list["tog_pulse"])
+		auto_pulse = !auto_pulse
+	else if(href_list["eject_cell"])
+		if(C)
+			C.forceMove(get_turf(src))
+			C = null
+			usr << browse(null, "window=DSES")
+			return
+	updateSelfDialog(usr)
+
+/obj/item/device/dses/updateSelfDialog(mob/user)
+	src.attack_self(user)
+
+/obj/item/device/dses/process()
+	if(auto_pulse && (!last_pulse || last_pulse <= world.time+pulse_between))
+		handle_pulse()
+
+/obj/item/device/dses/proc/handle_pulse(mob/user)
+	if(!last_pulse || last_pulse <= world.time+pulse_between)
+		var/atom/detected
+		detected = pulse()
+
+		if(!detected)
+			if(user)
+				to_chat(user, "<span class = 'warning'>No structures detected.</span>")
+			playsound(src, 'sound/machines/buzz-sigh.ogg', 50, 1, gas_modified = 0)
+			return
+
+		playsound(src, 'sound/machines/ping.ogg', 50, 1, gas_modified = 0)
+
+		if(module_in_list("dirGET"))
+			var/detect_dir = get_dir(src, detected)
+			switch(detect_dir)
+				if(1) last_direction = "North"
+				if(2) last_direction = "South"
+				if(4) last_direction = "East"
+				if(5) last_direction = "North East"
+				if(6) last_direction = "South East"
+				if(8) last_direction = "West"
+				if(9) last_direction = "North West"
+				if(10) last_direction = "South West"
+
+			if(user)
+				to_chat(user, last_direction)
+
+		if(module_in_list("crdLOG"))
+			if(detected in positive_locations)
+				positive_locations -= detected
+			positive_locations += detected
+
+		if(module_in_list("getDST"))
+			last_distance = get_dist(get_turf(src), get_turf(detected))
+
+/obj/item/device/dses/proc/pulse()
+	var/list/can_see = range(pulse_range,get_turf(src))
```

Comment:
Use `trange` here.

Actually `spiral_range` might be fun but that might not perform very well.